server/src/test/java/org/elasticsearch/index/translog/TranslogTests.java

```diff
@@ -1507,6 +1512,193 @@ public void testSnapshotFromStreamInput() throws IOException {
+
+            final Set<Long> expectedSeqNoSet = new HashSet<>(expectedSeqNo);
+            try (Translog.Snapshot snapshot = translog.newSnapshot()) {
+                Translog.Operation next;
+                while ((next = snapshot.next()) != null) {
+                    if (expectedForAttempt.contains(next.seqNo())) {
+                        assertThat(next.primaryTerm(), is(rollover ? primaryTerm.get() - 1 : primaryTerm.get()));
+                    }
+                    final boolean remove = expectedSeqNoSet.remove(next.seqNo());
+                    assertThat(next.primaryTerm() + ":" + next.seqNo() + " is expected", remove, is(true));
+                }
+                // duplicates with lower primary term make skipped operation even higher (+ there is overlapping of them)
+                assertThat(snapshot.skippedOperations(), greaterThanOrEqualTo(skipped));
+            }
+            assertThat("expectedSeqNoSet:" + expectedSeqNoSet, expectedSeqNoSet, hasSize(0));
+        }
+    }
+
+    public void testRandomExceptionsOnTrimOperations( ) throws Exception {
+        Path tempDir = createTempDir();
+        final FailSwitch fail = new FailSwitch();
+        fail.failNever();
+        TranslogConfig config = getTranslogConfig(tempDir);
+        List<FileChannel> fileChannels = new ArrayList<>();
+        final Translog failableTLog =
+            getFailableTranslog(fail, config, randomBoolean(), false, null, createTranslogDeletionPolicy(), fileChannels);
+
+        Throwable expectedException = null;
+        int translogOperations = 0;
+        for(int attempt = 0; attempt < 10; attempt++) {
+            int maxTrimmedSeqNo;
+            fail.failNever();
+            int extraTranslogOperations = randomIntBetween(10, 100);
+
+            List<Integer> ops = IntStream.range(translogOperations, translogOperations + extraTranslogOperations)
+                .boxed().collect(Collectors.toList());
+            Randomness.shuffle(ops);
+            for (int op : ops) {
+                String ascii = randomAlphaOfLengthBetween(1, 50);
+                Translog.Index operation = new Translog.Index("test", "" + op, op,
+                    primaryTerm.get(), ascii.getBytes("UTF-8"));
+
+                failableTLog.add(operation);
+            }
+
+            translogOperations += extraTranslogOperations;
+
+            // at least one roll + inc of primary term has to be there - otherwise trim would not take place at all
+            boolean rollover = attempt == 0 || randomBoolean();
+            if (rollover) {
+                primaryTerm.incrementAndGet();
+                failableTLog.rollGeneration();
+            }
+
+            maxTrimmedSeqNo = rollover ? translogOperations - randomIntBetween(4, 8) : translogOperations + 1;
+
+            // chance to fail is proportional to number of attempt to get on the last attempt 100% chance of failure
+            fail.failRate(10 + 10 * attempt);
```

Comment:
This is too subtle - if someone reduces the number of iterations it is very easy to miss the point that it has implications for this. How about making a max attempt constant (set it to 5) and then here do something like fail.failRate(attempt == MAX_ATTEMPT ? 100 : 30)